### ISSUE TITLE
SoFiA crashes onPawsey when using reliability

### ISSUE THREAD

**Issue by apopping** [SoFiA-Admin/SoFiA#101]
SoFiA is running on Pawsey, however crashes when using the reliability calculations.
In a previous telecon it was sugggested that this has to do with the covariance matrix.

In the parameter file I have tried both
reliability.usecov = true
and 
reliability.usecov = false

but I get a similar error:

--- SoFiA: Determining reliability ---
Application 5644800 exit codes: 1
Application 5644800 resources: utime ~8s, stime ~2s, Rss ~149072, inblocks ~65125, outblo
cks ~57

**Comment by paoloserra** [comment]
Have you tried with a higher source-finding threshold to see whether the error depends on the nr of sources detected?

**Comment by SoFiA-Admin** [comment]
I just downloaded the latest SoFiA code, and SoFiA also crashes on my machine when using reliability. Unlike on Pawsey, I actually get a detailed error message:

    --- SoFiA: Determining reliability ---
    Traceback (most recent call last):
      File "~/SoFiA/git/sofia_pipeline.py", line 329, in <module>
        objects,reliable = addrel.EstimateRel(np.array(objects), outroot, catParNames, **Parameters['reliability'])
      File "~/build/lib.linux-x86_64-2.7/sofia/addrel.py", line 41, in EstimateRel
        for ii in range(0,len(parSpace)): parCol.append(parNames.index(parSpace[ii]))
    ValueError: tuple.index(x): x not in tuple
    Pipeline failed with exit code 1.

Does any of these messages ring a bell?

**Comment by RussellJurek** [comment]
So it compiled fine with your version of gsl?

Cheers,
Russell

**Comment by SoFiA-Admin** [comment]
Yes, your gsl fix seems to work, and I get realistic BF fitting results.

**Comment by paoloserra** [comment]
Can you replace
for ii in range(0,len(parSpace)): parCol.append(parNames.index(parSpace[ii]))

with
for ii in range(0,len(parSpace)):
  print parSpace[ii]
  parCol.append(parNames.index(parSpace[ii]))

That should allow you to see what parameter is causing troubles.
I have run some tests incl the reliability calculation and cannot reproduce that error, but admittedly that will have been on different data and with a different parameter file.

**Comment by SoFiA-Admin** [comment]
The output is `ftot`.

**Comment by paoloserra** [comment]
Seems like you are using an old parameter file. "ftot" is the old name for the total flux parameter. The name was changed in https://github.com/SoFiA-Admin/SoFiA/commit/962174100ae19b5585b9fa10b3baf94654c40942 . Current default is
reliability.parSpace    =       ['n_pix','snr_sum','snr_max']
We should however implement some more useful error message.

**Comment by SoFiA-Admin** [comment]
Yep, that did the trick! Could the same problem be causing the crash on Pawsey, or are we dealing with a different problem here?

**Comment by apopping** [comment]
Pawsey is undergoing maintenance today so cannot try, but will test it tomorrow!

**Comment by paoloserra** [comment]
It seems that also on Magnus we were using an old parameter file, where:

reliability.parSpace = ['SNRsum','SNRmax','NRvox']

I'll try to add some checks in sofia_pipeline.py today.

**Comment by paoloserra** [comment]
OK, added some checks. See https://github.com/SoFiA-Admin/SoFiA/commit/06c8b6b76750300be2fe8cf7cae76e7ab0b6b19f . When using an illegal reliability.parSpace you should now receive a useful error message.

**Comment by SoFiA-Admin** [comment]
Thanks, Paolo, for fixing this so quickly.

**Comment by SoFiA-Admin** [comment]
By the way, this means that all of our old parameter files, e.g. the ones coming with the official test data cube or the SoFiA tutorial, will no longer work and will have to be replaced, too! Should we just update those, or rather provide two versions, one for SoFiA 0.x and another one for 1.x?

**Comment by paoloserra** [comment]
Old parameter files will work if they do not request the reliability calculation.

Since we are still at an early stage of the project, and we should encourage users to move to SoFiA v1.0, I would favor providing only parameter files compliant with SoFiA v1.0. That's all new users will need, while users of old SoFiA versions will already have the old parameter file.

**Comment by SoFiA-Admin** [comment]
> Old parameter files will work if they do not request the reliability calculation.

Unfortunately, the test data cube parameter file and one of the four tutorial parameter files do enable reliability calculation. I will therefore bring all of the files up-to-date to comply with the 1.0 requirements. I agree that it should be okay to only provide files for 1.0 and remove the 0.x files.

**Comment by RussellJurek** [comment]
It seems to me the best solution is a conversion function in the input module. The conversion function would map old parameters to their corresponding new definitions. That way it's handled in the background so users don't see it. If it fails to do this conversion, then it can quit SoFiA during this stage when a user hasn't invested time and resources in a run that will fail. It would also provide a natural place to generate more meaningful error messages.

**Comment by paoloserra** [comment]
I don't know, I think that may be a good solution for future changes, say post v1.0. But for now I'd assume that users would much rather use v1.0 than an older version.

**Comment by SoFiA-Admin** [comment]
Okay, I changed all the files (still to be uploaded to GitHub), and both the tutorial and test data set work fine and produce the expected results. I also implemented two additional changes in the process:

1. I replaced the **absolute path** of the input file with just the file name in all parameter files. As I just found out, SoFiA can actually handle relative paths, so if the data cube is in the same directory as the parameter file, it is sufficient to just specify the file name without any path. This means that users can run the parameter files out of the box without the need to first specify the location of the input cube relative to their local file system.
2. I disabled output file **compression** in the test data cube parameter file. There is no need to have this switched on, given the tiny size of the test data cube. Without compression it will be much easier for users to inspect the outcome straight away.

> It seems to me the best solution is a conversion function in the input module.

I’m not sure if that’s a desirable solution. This sounds a bit like the HTML quirks mode in web browsers: “We know that your HTML code is full of errors, but we will silently ignore them and render the page anyway.” This philosophy has resulted in 99.9...% of all HTML files on the internet to be non-compliant, because errors go unnoticed. I would rather prefer to point out to the user that the file is wrong and how to fix it.
